nvflare/job_config/fed_job.py

```diff
@@ -213,33 +250,36 @@ def to(
         self._validate_target(target)
 
         if isinstance(obj, Controller):
-            if target not in self._deploy_map:
-                self._deploy_map[target] = ControllerApp(key_metric=self.key_metric)
-            self._deploy_map[target].add_controller(obj, id)
-        elif isinstance(obj, Executor):
+            if target != "server":  # add client-side controllers as components
+                if target not in self._deploy_map:
+                    raise ValueError(
+                        f"{target} doesn't have an `Executor`. Deploy one first before adding client-side controllers!"
+                    )
+                self._deploy_map[target].add_component(obj, id)
+            else:
+                if target not in self._deploy_map:
+                    self._deploy_map[target] = ControllerApp(key_metric=self.key_metric)
+                self._deploy_map[target].add_controller(obj, id)
+        elif isinstance(obj, (Executor, ScriptExecutor)):
+            if target == "server":
+                raise ValueError(
+                    f"`Executor` must be assigned to a client, but tried to assign `Executor` {obj} to 'server'!"
+                )
             if target not in self._deploy_map:
                 self._deploy_map[target] = ExecutorApp()
+
             if isinstance(obj, ScriptExecutor):
-                external_scripts = [obj._task_script_path]
-                self._deploy_map[target].add_external_scripts(external_scripts)
+                self._add_script_executor(obj, target, tasks)
+            else:
+                self._deploy_map[target].add_executor(obj, tasks=tasks)
```

Comment:
Shouldn't this be obj.base_obj?

Reply:
obj.base_obj is already passed into _to() from to()